python docs: hooks, offline, get instance, pollingmode (#200)
configcat/docs: website/docs/sdk-reference/python.md

| `flag_overrides`          | Local feature flag & setting overrides. [More about feature flag overrides](#flag-overrides)                                | None    |
| `data_governance`         | Describes the location of your feature flag and setting data within the ConfigCat CDN. This parameter needs to be in sync with your Data Governance preferences. [More about Data Governance](advanced/data-governance.md). Available options: `Global`, `EuOnly`. | `Global`|
| `default_user`            | Sets the default user. [More about default user](#default-user). | None |
| `hooks`                   | Used to subscribe events that the SDK sends in specific scenarios. [More about hooks](#hooks). | None |
| `offline`                 | Indicates whether the SDK should be initialized in offline mode or not. [More about offline mode.](#online--offline-mode). | False |

```python
client = configcatclient.get('#YOUR-SDK-KEY#',
    ConfigCatOptions(
        polling_mode=PollingMode.auto_poll()
    )
)
```

:::caution
We strongly recommend you to use the _ConfigCat Client_ as a Singleton object in your application.

> The `configcatclient.get("#YOUR-SDK-KEY#")` static factory method constructs singleton client instances for your SDK keys.

Reviewer: ```suggestion
The `configcatclient.get()` static factory method constructs singleton client instances for your SDK keys.
```

| `user`          | Optional, _User Object_. Essential when using Targeting. [Read more about Targeting.](advanced/targeting.md) |

```python
value = client.get_value(
    'keyOfMySetting', # Setting Key
    False, # Default value

>     User('435170f4-8a8b-4b67-a723-505ac7cdea92') # Optional User Object

Reviewer: Please consider https://github.com/configcat/docs/pull/206

)
```

## Anatomy of `get_value_details()`

`get_value_details()` is similar to `get_value()` but instead of returning the evaluated value only, it gives an _EvaluationDetails_ 
object with more detailed information about the evaluation result.

| Parameters     | Description                                                                                                  |
| -------------- | ------------------------------------------------------------------------------------------------------------ |
| `key`          | **REQUIRED.** Setting-specific key. Set on _ConfigCat Dashboard_ for each setting.                           |
| `defaultValue` | **REQUIRED.** This value will be returned in case of an error.                                               |
| `user`         | Optional, _User Object_. Essential when using Targeting. [Read more about Targeting.](advanced/targeting.md) |

```python
details = client.get_value_details(
    'keyOfMySetting', # Setting Key
    False, # Default value

>     User('435170f4-8a8b-4b67-a723-505ac7cdea92') # Optional User Object

Reviewer: also https://github.com/configcat/docs/pull/206

user_object = User('john@example.com')
```

### Customized user object creation

| Parameters   | Description                                                                                                                     |
| ------------ | ------------------------------------------------------------------------------------------------------------------------------- |
| `identifier` | **REQUIRED.** Unique identifier of a user in your application. Can be any value, even an email address.                         |
| `email`      | Optional parameter for easier targeting rule definitions.                                                                       |
| `country`    | Optional parameter for easier targeting rule definitions.                                                                       |
| `custom`     | Optional dictionary for custom attributes of a user for advanced targeting rule definitions. e.g. User role, Subscription type. |

```python
user_object = User(

>   '435170f4-8a8b-4b67-a723-505ac7cdea92',

Reviewer: also https://github.com/configcat/docs/pull/206

    )
)
```

Available options:

| Option Parameter             | Description                  | Default |
| ---------------------------- | ---------------------------- | ------- |
| `cache_time_to_live_seconds` | Cache TTL.                   | 60      |

### Manual polling

Manual polling gives you full control over when the `config.json` (with the setting values) is downloaded. _ConfigCat SDK_ will not update them automatically. Calling `force_refresh()` is your application's responsibility.

```python
client = configcatclient.get('#YOUR-SDK-KEY#',
    ConfigCatOptions(
        polling_mode=PollingMode.manual_poll()
    )
)
client.force_refresh()
```

> `get_value()` returns `default_value` if the cache is empty. Call `force_refresh()` to update the cache.

```python
client = configcatclient.get('#YOUR-SDK-KEY#',
    ConfigCatOptions(
        polling_mode=PollingMode.manual_poll()
    )
)
value = client.get_value('key', 'my default value') # Returns "my default value"
client.force_refresh()
value = client.get_value('key', 'my default value') # Returns "value from server"
```

## Hooks

With the following hooks you can subscribe to particular events fired by the SDK:

> - `on_client_ready()`: This event is sent when the SDK reaches the ready state. If the SDK is configured with lazy load or manual polling it's considered ready right after instantiation.

Reviewer: ```suggestion
- `on_client_ready()`: This event is sent when the SDK reaches the ready state. If the SDK is set up with lazy load or manual polling it's considered ready right after instantiation.
```

If it's using auto polling, the ready state is reached when the SDK has a valid config.json loaded into memory either from cache or from HTTP. If the config couldn't be loaded neither from cache nor from HTTP the `on_client_ready` event fires when the auto polling's `max_init_wait_time_seconds` is reached.

- `on_config_changed(config: dict)`: This event is sent when the SDK loads a valid config.json into memory from cache, and each subsequent time when the loaded config.json changes via HTTP.

- `on_flag_evaluated(evaluation_details: EvaluationDetails)`: This event is sent each time when the SDK evaluates a feature flag or setting. The event sends the same evaluation details that you would get from [`get_value_details()`](#anatomy-of-getvaluedetails).

- `on_error(error: str)`: This event is sent when an error occurs within the ConfigCat SDK.

You can subscribe to these events either on SDK initialization: 
```python
def on_flag_evaluated(evaluation_details):
    # handle the event
    pass

client = configcatclient.get('#YOUR-SDK-KEY#',
    ConfigCatOptions(
        hooks=Hooks(on_flag_evaluated=on_flag_evaluated)
    )
)
```

> or with the `get_hooks()` property of the ConfigCat client:

Reviewer: im not sure, but this seems like a method not a prop, sorry if i misunderstood.

| `user`     | Optional, _User Object_. Essential when using Targeting. [Read more about Targeting.](advanced/targeting.md) |

```python
client = configcatclient.get('#YOUR-SDK-KEY#')
all_values = client.get_all_values(User('435170f4-8a8b-4b67-a723-505ac7cdea92'))  # Optional User Object
```

## Custom cache

You have the option to inject your custom cache implementation into the client. All you have to do is to inherit from the ConfigCache abstract class:

```python
from configcatclient.interfaces import ConfigCache

class InMemoryConfigCache(ConfigCache):
    def __init__(self):
        self._value = {}

    def get(self, key):

>         # here you have to return with the cached value

Reviewer: ```suggestion
        # you should return the cached value
```

        return self._value.get(key)

    def set(self, key, value):

>         # here you have to store the new value in the cache

Reviewer: ```suggestion
        # you should cache the new value
```

| `user`     | Optional, _User Object_. Essential when using Targeting. [Read more about Targeting.](advanced/targeting.md) |

```python
client = configcatclient.get('#YOUR-SDK-KEY#')

> all_values = client.get_all_values(User('435170f4-8a8b-4b67-a723-505ac7cdea92'))  # Optional User Object

Reviewer: also https://github.com/configcat/docs/pull/206

```

## `get_all_value_details()`

Evaluates and returns the detailed values of all feature flags and settings. Passing a [User Object](#user-object) is optional.

```python
client = configcatclient.get('#YOUR-SDK-KEY#')

> all_value_details = client.get_all_value_details(User('435170f4-8a8b-4b67-a723-505ac7cdea92'))  # Optional User Object

Reviewer: also https://github.com/configcat/docs/pull/206
